hibernate-osgi/src/main/java/org/hibernate/osgi/OsgiClassLoader.java

```diff
@@ -29,12 +32,18 @@
 public class OsgiClassLoader extends ClassLoader implements Stoppable {
 	// Leave these as Sets -- addClassLoader or addBundle may be called more
 	// than once if a SF or EMF is closed and re-created.
-	private Set<ClassLoader> classLoaders = new LinkedHashSet<ClassLoader>();
-	private Set<Bundle> bundles = new LinkedHashSet<Bundle>();
+	// HHH-12553: must be thread-safe. Concurrent impl. would be best, but we have to retain insertion-order.
+	private Set<ClassLoader> classLoaders = Collections.synchronizedSet(new LinkedHashSet<ClassLoader>());
+	private Set<Bundle> bundles = Collections.synchronizedSet(new LinkedHashSet<Bundle>());
+
+	private ConcurrentMap<String, Class<?>> classCache = new ConcurrentHashMap<String, Class<?>>();
+	private ConcurrentMap<String, URL> resourceCache = new ConcurrentHashMap<String, URL>();
+
+	static
+	{
+		ClassLoader.registerAsParallelCapable();
```

Comment:
Why is this change needed since the issue is about preventing the `ConcurrentModificationException`?

Reply:
It is not needed to fix the CMFEx. But since the class became thread-safe with the change, I think it makes sense to mark is as such.
See HHH-12553 for some more comments.